src/fs/rock/RockIoState.cc

```diff
@@ -245,44 +255,38 @@ Rock::IoState::writeToBuffer(char const *buf, size_t size)
 }
 
 /// write what was buffered during write() calls
-/// negative sidNext means this is the last write request for this entry
 void
-Rock::IoState::writeToDisk(const SlotId sidNextProposal)
+Rock::IoState::writeToDisk()
 {
     assert(theFile != NULL);
     assert(theBuf.size >= sizeof(DbCellHeader));
 
-    const bool lastWrite = sidNextProposal < 0;
+    assert((sidFirst < 0) == (sidCurrent < 0));
+    if (sidFirst < 0) // this is the first disk write
+        sidCurrent = sidFirst = dir->reserveSlotForWriting();
+
+    // negative sidNext means this is the last write request for this entry
+    const bool lastWrite = sidNext < 0;
+    // here, eof means that we are writing the right-most entry slot
     const bool eof = lastWrite &&
                      // either not updating or the updating reader has loaded everything
                      (touchingStoreEntry() || staleSplicingPointNext < 0);
-    // approve sidNextProposal unless _updating_ the last slot
-    const SlotId sidNext = (!touchingStoreEntry() && lastWrite) ?
-                           staleSplicingPointNext : sidNextProposal;
-    debugs(79, 5, "sidNext:" << sidNextProposal << "=>" << sidNext << " eof=" << eof);
+    debugs(79, 5, "sidCurrent=" << sidCurrent << " sidNext=" << sidNext << " eof=" << eof);
 
     // TODO: if DiskIO module is mmap-based, we should be writing whole pages
     // to avoid triggering read-page;new_head+old_tail;write-page overheads
 
-    writeBufToDisk(sidNext, eof, lastWrite);
-    theBuf.clear();
-
-    sidCurrent = sidNext;
-}
-
-/// creates and submits a request to write current slot buffer to disk
-/// eof is true if and only this is the last slot
-void
-Rock::IoState::writeBufToDisk(const SlotId sidNext, const bool eof, const bool lastWrite)
-{
-    // no slots after the last/eof slot (but partial slots may have a nil next)
-    assert(!eof || sidNext < 0);
+    assert(eof == lastWrite); // no slots after eof
```

Comment:
> below we require that 'eof' and 'lastWrite' mean the same.

Where do we do that? AFAICT, the two variables have different meaning and, hence, will have different values in some cases. If you are confused by the assignment quoted below, please note that `r->eof` does not mean the same as local `eof` (unfortunately). In Rock context, `r->eof` means what lastWrite means locally.

```C++
r->eof = lastWrite;
```

We could, and probably should, rename both `eof`s but that renaming is outside this PR scope and will obscure this PR changes.

If you see any code that contradicts my statements, please let me know. Otherwise, please fix/revert this assertion.

Reply:
>  will have different values in some cases

How could they (i.e., eof and lastWrite) have different values, if we assert that they cannot?:

```
const bool lastWrite = sidNext < 0;
[...]
assert(eof == (sidNext < 0)); 
```
Also I cannot understand what is wrong in rewriting the above assertion into `assert(eof == lastWrite)`.
```

Reply:
> How could they (i.e., eof and lastWrite) have different values, if we assert that they cannot?

Strange question. Their values are determined by their initialization code, not by assertions. If initialization does not match assertion, then we need to fix that bug by changing initialization/description and/or assertions.

> Also I cannot understand what is wrong in rewriting the above assertion into `assert(eof == lastWrite)`.

The two variables are described to mean two different things and are initialized using different code. If their values are always the same, then something went wrong somewhere.

Reply:
AFAICT, the following equality assertion is wrong:

```C++
assert(eof == (sidNext < 0)); // no slots after eof
```

And should be replaced with this implication:

```C++
assert(!eof || (sidNext < 0)); // no slots after eof
```

The comment (i.e. the intended assertion meaning) is and was correct.

Reply:
And that correct/implication assertion was already in [master](https://github.com/squid-cache/squid/blob/424f369/src/fs/rock/RockIoState.cc#L278). Its description was stale, but I should not have changed the assertion itself when updating the description. That change was wrong.

Reply:
Done.